2단계 미션 제출합니다.

HyewonKkang opened the pull request:
안녕하세요! 1단계에서 받은 피드백을 포함해서 3단계를 위한 리팩토링도 조금 진행해보았습니다.
많은 피드백 주시면 감사하겠습니다 :)

adeldel commented on src/main/java/request/RequestUri.java:
```diff
@@ -0,0 +1,74 @@
+        return new RequestUri(getMethodType(splitLine[METHOD_INDEX_IN_LINE])
+                , extractPath(splitLine[URI_INDEX_IN_LINE])
+                , new HashMap<>());
+    }
```
`splitLine[URI_INDEX_IN_LINE] `이 세번이나 반복되고 있습니다. 변수로 선언해서 해당 변수를 사용하는건 어떨까요?
마찬가지로 메소드도요.
```
String requestURI = splitLine[URI_INDEX_IN_LINE];
String requestMethod = splitLine[METHOD_INDEX_IN_LINE];

adeldel commented on src/main/java/request/RequestUri.java:
```diff
@@ -0,0 +1,74 @@
+            return RequestMethod.PATCH;
+        }
+        throw new MethodNotFoundException();
+    }
```
enum의 `valueOf` 는 내부적으로 자기 자신 instance를 반환합니다. 
존재하지 않는 값일때는 exception을 던지므로 RequestMethod에는 현재 지원하는 메소드만 지정해 놓고 valueOf를 사용하는건  어떨까요?
```suggestion
private static RequestMethod getMethodType(String line) {
        String method = line.split(" ")[METHOD_INDEX_IN_LINE];
        try {
            return RequestMethod.valueOf(method);
        } catch (IllegalArgumentException ex) {
            throw new MethodNotFoundException();
        }
}

adeldel commented on src/main/java/utils/HttpRequestUtils.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public static Optional<String> extractParams(String uri) {
+        String[] tmp = uri.split(PARAMETER_START_REGEX);
```
tmp 보다는 의미있는 변수명을 사용해주세요.

adeldel commented on src/main/java/utils/HttpRequestUtils.java:
```diff
@@ -0,0 +1,49 @@
+        return result;
+    }
+
+    private static void isInvalidRequest(String[] splitPair) {
```
소소하지만 `isInvalidRequest`  메소드명보다  조금 더 validation 을 기대하는 이름으로 변경하면 좋겠습니다. 
(메소드명은 짓기 나름이지만 보통 is prefix는 true/false 값을 기대하는 느낌입니다.)

adeldel commented on src/main/java/webserver/ControllerHandler.java:
```diff
@@ -0,0 +1,25 @@
+        controllerMap.put("/user/login", new LoginController());
+        controllerMap.put("/user/logout", new LogoutController());
+        controllerMap.put("/user/list.html", new ListUserController());
+    }
```
생성자도 괜찮지만 static 블럭 (initialization block)으로 미리 만들어두면 좋을것 같아요.
```
static {
        controllerMap.put("/user/create", new CreateUserController());
        controllerMap.put("/user/login", new LoginController());
        controllerMap.put("/user/logout", new LogoutController());
        controllerMap.put("/user/list.html", new ListUserController());
}

adeldel commented:
안녕하세요. hendo 😄  마지막까지 미션하시느라 수고가 많으십니다.
질문주신 내용에 대한 개인적인 의견을 드려볼께요.

Q. 코드를 짜다보니 Response의 경우 객체의 책임과 테스트 코드 작성 두 부분 모두에 있어서 굉장히 애매하다는 생각이 들었습니다.
이부분에 대해서 조언을 주실 수 있을까요?
A. 현재 만들어주신 HttpResponse 객체의 책임부분은 어느 정도 잘 분담되었다고 보여지구요.
테스트코드에 대해서는 방법을 제시해보자면 🤔   FileOutputStream 을 이용하여 내가 만들고자 하는 response를 직접 파일로 써볼 수 있을것 같아요.  음.. 이런식으로 응답을 직접 outputstream으로 써보는 테스트정도는 해볼 수 있겠네요. 
````
public class HttpResponseTest {
	private String testResources = "./src/test/resources/";

	@Test
	public void response302HeaderTest() throws Exception {
		HttpResponse response = HttpResponse.from(createOutputStream("response302Header.txt"));
		response.response302Header("/user/login.html", "false");
	}

	private OutputStream createOutputStream(String filename) throws FileNotFoundException {
		return new FileOutputStream(new File(testResources + filename));
	}
}
````
Q. 패키지의 구분을에 대해서 조금 더 조언을 얻고싶습니다..!
지금 제가 만든 코드 중 request나 response는 dto에 해당할까요..?
로직을 처리하는 부분(?)이 상대적으로 많아 도메인 처럼 느껴지기도하고, 이건 was니까 도메인보다는 dto같기두 하고..
애매한 부분이 많아서 그냥 request 패키지와 response 패키지로 나누었는데, 이부분에 대해 질문 드리고싶습니다!
A. 구조적으로 크게 문제되는건 없어보여요. 만들어주신 request/response 패키지도 이상하지는 않지만, 저는 해당 클래스들이 domain의 역할에 가깝다고 생각됩니다.  HTTP Request 요청에 포함되는 정보를 충분히 역할분담해주셨다고 생각됩니다. 
그에 따른 유닛테스트도 잘 만들어주셨구요. 답변이 되었을까요 😅

adeldel reviewed (changes requested):
안녕하세요 hendo! http 웹서버 리뷰를 맡은 아델이라고 합니다. 😁
전반적으로 기능구현을 잘 해주셨고, 구조적으로 큰 문제는 없어보여요.
크게 변경할 것은 없지만, 조금 더 신경쓰면 좋을 포인트에 댓글을 달아두었습니다.
참고로 질문주신 내용도 코멘트로 답변드렸으니 확인부탁드릴께요.
수정반영 후 재요청 부탁드립니다. 화이팅 💪🏾

Chaeoon-Park commented:
안녕하세요 adeldel!
우선 코드를 너무 꼼꼼히 읽어주시고 리뷰를 작성해주신것 같아서 감사하다는 말씀 먼저 드리고 싶습니다..!!!!
말씀해주신 피드를 모두 적용하고자 노력했는데, 잘 적용이 되었을지 모르겠네요.. ! 
질문에 대한 답변도 매우 도움이 되었습니다!! 감사합니다.

adeldel이 알려주신 리팩토링을 진행하다보니, 패턴을 잘 적용하면 split이 자주 일어나는 코드가 훨씬 나아질 수도있겠다는 생각이 들었는데,
고비용(?)의 리팩토링을 진행하다가 자칫하면 금요일까지 step3까지 리뷰를 받지 못할까봐 일단 리뷰요청 먼저 드리게 되었습니다..!
너무 감사하고, 잘 부탁 드리겠습니다..!ㅎㅎ

마지막으로 질문을 조금... 아니 많이..? 드리자면..!

1. 요청 하신것 이외에도 RequestHandler의 run부분에도 커스텀 에러를 추가해 보았는데요
`catch (IOException e) {
            logger.error(e.getMessage());
            throw new ResponseCreateFailException();
        }`
사실 이전에 리퀘스트를 생성하면서 충분히 다른 에러에 의해 걸러질 것이라고 생각되기는 하는데, 
트라이 캐치문으로 에러가 발생하면 에러를 그래도 추가해주는것이 좋은지 여쭤보고싶습니다..!

2. 서비스 레이어 분리에 대해 질문드리고 싶은 점이 있는데요..!
httpRequest와 httpResponse가 도메인 로직을 포함하고 있기는 하지만, 컨트롤러를 돌아다니는 것은 피할 수 없을 것 같아서
최소한 서비스로는 가지 않도록 하였는데 아무래도 도메인 로직을 포함한 부분과 컨트롤러에서 쓰이는 부분을 분리해야할까요..?
httpRequest를 만드는 과정에서 도메인로직을 피할 수는 없을 것 같다는 생각도 들고요..! 
httpResponse의 경우 쓰는부분과 도메인 로직 부분으로 생각되는 부분을 분리해도, 굳이란 생각이 조금 드는거 같기두 하고요..!
혹시 어떻게 생각하시는지 자문을 구할 수 있을까요? (아직은 제가 똑바로 이해하지 못한것일까요 ㅠㅠ)

3. 정규표현식으로 말씀하신대로 유효성 검사를 하는 과정에서 생각난 것인데요..!
'     
private static final String REQUEST_LINE_PATTERN = "\\S+ /\\S+ HTTP/1.1";
private static void validateRequestLine(String line) {
        if (!Pattern.matches(REQUEST_LINE_PATTERN, line)) {
            throw new InvalidRequestLineException();
        }
    }'

저는 이미 http메소드가 존재하는지, 아니면 파라미터의 값이 null로 오지는 않았는지 등을 다른 메서드에서 활용하고 있기 때문에,
REQUEST_LINE_PATTERN을 저정도로 끝냈는데, 정규 표현식을 공부하다 보니, 정규표현식을 빡빡하게 만들면 한번에 검사할 수 있을 것이라는 생각이 들더라고요..!
보통은 정규표현식을 조금 더 빡빡하게 만들어서 한번에 검사하는 편인가요?
정규표현식을 처음 써보았는데, 빡빡하게 만들면 많은 부분을 검사할 수 있을 것 같다는 생각과 동시에, 가독성이 너무 떨어져서 꼭 필요한 부분이나 간단한 형식 이외에는 쓰기 싫다는 생각도 조금 들더라고요..? 현업에서는 정규표현식이 적극(?) 활용되는지 궁금해요..!


쓰다보니 질문 투성이네요 ㅠㅠㅠ 바쁘실텐데 정성스러운 리뷰 항상 감사합니다..! ㅠㅠ

adeldel commented:
안녕하세요 hendo ! 질문주신 사항에 대해 고민해보고 제 경험에 빗대어 답변을 드립니다. 

> Q. 트라이 캐치문으로 에러가 발생하면 에러를 그래도 추가해주는것이 좋은지 여쭤보고싶습니다..!
A. 무시해도 좋을 에러라면 catch해서 로그를 남겨서 추후 이런 상황에서 에러가 발생하는구나라고 개발자가 파악하기 좋을것이며, catch해서 throw하는 경우에는 실제 유저에게 비정상적인 상황에 대해 알려줄 수 있습니다. 다만, 의도치 않은 에러를 클라이언트에게 전달하면 안되겠죠. 개발을 하다보면 exception을 자칫 처리 안했다가 (흔히 에러를 먹는다고 하죠) 문제 해결할때 힘든 경우가 있습니다.
그만큼 에러처리는 중요한 포인트인것 같아요. 
[참고자료를](https://www.nextree.co.kr/p3239/) 보시고 예외처리에 대해 찾아보시면 좀더 이해가 되실 겁니다. 

> Q. 혹시 어떻게 생각하시는지 자문을 구할 수 있을까요?
A. 사실 현업을 하다보면 말씀하신 서비스와 도메인의 경계가 모호해질때가 있는데요. 객체지향적인 방식은 도메인에 좀 더 치중해서 기능을 위임해야 합니다. 서비스레이어는 가능한 가볍게 구현하고, 여러 Service나 Dao을 연결하는 역할이나 트랜잭션 관리 등의 역할이 맞다고 생각합니다. 질문에 대해 답변이 되었을런지 모르겠습니다. 😅 

> Q. 정규표현식으로 말씀하신대로 유효성 검사를 하는 과정에서 생각난 것인데요..!
.. 현업에서는 정규표현식이 적극(?) 활용되는지 궁금해요.
A. 정규식에 대해 말씀드리자면, (서비스 특성이 있겠지만) 저희조직은 아주 많이*100 사용하고 있습니다.
제가 맡고 있는 서비스가 오래된(20년) 서비스이다 보니 시간이 흐르면서, 지원해주거나 종료해야 하는 데이터가 많아서 이기도 하구요.
한 예로 가장 최근에 적용한 PC HTTPS 적용하면서, 예전에는 http로 된 이미지가 컨텐츠에 많이 있었는데 https 환경에서는 액박이미지로 뜨기 때문에 해당 이미지들을 모두 https로 변경해서 읽도록 해줘야 할때 html로 이뤄진 본문내에서 img태그를 찾아서 url을 교체해준다거나 이런 작업들이 이루어졌어요. 그럴때 정규표현식 regex가 적극적으로 사용되지요.
다만, 예전 경험상 regex를 잘못 사용했을때 사이드이펙트가 크고, 특히 Pattern.DOTALL 을 예전에 잘못 써서 변환시 속도가 아주아주 느려져서 서비스장애로 이어졌던 경험이 있어요. 그만큼 유용하지만 조심히 써야 하는게 패턴인듯 합니다.
자바에서 지원하는 String replace, replaceAll, replaceFirst 같은 메소드도 모두 regex를 지원하니, 단순한 치환은 Pattern을 만들어서 사용하는 것보다 낫지 않나 싶네요.

adeldel reviewed (approved):
안녕하세요 hendo 😄 
리뷰 반영을 아주 잘 해주셨고요. 기능상으로 정상 동작함을 확인했습니다.
질문주신 사항은 코멘트로 정리해서 남겨드렸습니다. (제 경험을 빗대어 답변드린거라 정답은 아니지만, ) 참고해주세요. 
이번 스탭은 여기서 머지할께요. 정말 수고많으셨습니다. 👍

HyewonKkang committed:
refactor: step1 코드 리뷰 반영

- 상수명 수정
- 함수명 수정
- `QueryStringParser` -> `RequestBodyParser`
- 테스트 코드 수정
- 중복 로직 분리

HyewonKkang committed:
refactor: `CustomResponse`로 분리

HyewonKkang committed:
refactor: `Response` body와 header로 분리

HyewonKkang committed:
refactor: `Controller` 인터페이스 및 추상클래스 생성, `SignUpController` 정의

HyewonKkang committed:
refactor: `RequestDispatcher`로 controller 로직 호출되도록 수정

HyewonKkang committed:
docs(README): 2단계 기능목록 추가

HyewonKkang committed:
feat: 로그인 판별 및 응답 기능

HyewonKkang committed:
refactor: `CustomResponse` 구조 리팩토링 및 body 클래스 제거

HyewonKkang committed:
feat: 쿠키 기반 로그인 기능

HyewonKkang committed:
feat: 로그인 쿠키 구현

HyewonKkang committed:
Merge remote-tracking branch 'origin/step2' into step2

HyewonKkang committed:
feat: 템플릿 엔진 적용 및 Session 구현체 추가

HyewonKkang committed:
feat: 로그인 상태에서 로그인페이지 접근 시 redirect 처리 및 User 객체 세션 값으로 저장

HyewonKkang committed:
feat: Cookie path 설정

HyewonKkang committed:
test: 테스트 코드 추가

shared-moon commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,8 @@
+import webserver.request.CustomRequest;
+import webserver.response.CustomResponse;
+
+public interface Controller {
```
이 인터페이스의 의미가 좀 없는 것 같아요, 어차피 추상클래스를 사용하고 있어서 굳이 필요 없는 부분 아닐까요 ?

shared-moon commented on src/main/java/controller/AbstractController.java:
```diff
@@ -0,0 +1,24 @@
+        return request.isMethodEqual(CustomMethod.POST);
+    }
+
+    protected abstract void doGet(CustomRequest request, CustomResponse response) throws Exception;
```
요 구조로 가면 하나의 패스에 get/post가 묶여서 만들어져야해서 쓸모없는 메서드나 구현상에 제약이 생길 것 같아요. 메서드 타입에 제약을 두지 말고 패스마다 메서드를 따로 연결 할 수 있는 구조는 어떻게 생각하시나요 ?

shared-moon replied:
약간 헷갈리실 것 같아서 .. 지금도 패스에 객체를 연결하고 계시긴 한데, 클래스가 옛날 서블릿처럼 조금 유연하지 못한 구조인 것 같아서요. Map을 <String, Function<Request,Response>> 처럼 만들어서 해당하는 메서드를 직접 맵핑하면 더 자유로운 구조가 되지 않을까 싶었어요 ~ 3단계 미션에서 한번 시도해보면 좋을 것 같아요!

shared-moon commented on src/main/java/controller/LoginController.java:
```diff
@@ -0,0 +1,55 @@
+public class LoginController extends AbstractController {
+
+    @Override
+    protected void doGet(CustomRequest request, CustomResponse response) throws Exception {
```
요 요청은 언제 들어오는거에요 ?

shared-moon commented on src/main/java/webserver/dispatcher/RequestDispatcher.java:
```diff
@@ -0,0 +1,33 @@
+        if (controller != null) {
+            return controller;
+        }
+        return new DefaultController();
```
요거 DefaultController는 Get 요청에서 리소스만 보고 있는데, 위에 걸리는 게 없을경우 이게 폴백으로 나가면 Post요청은 그냥 무시되지 않을까요 ?

Default로 처리 가능한녀석인지를 보고 안될경우 404를 내려주는것도 좋을 것 같아요 ~

shared-moon commented on src/main/java/webserver/request/CustomHeaders.java:
```diff
@@ -22,6 +24,14 @@ private Map<String, String> parseHeaders(final List<String> headersInput) {
 
+    public Cookie getCookie() {
+        return Cookie.from(elements.entrySet().stream()
+                .filter(entry -> entry.getKey().equals("Cookie"))
```
Cookie 상수로 정의되면 좋을 것 같아요 ~

shared-moon commented on src/main/java/webserver/response/CustomResponse.java:
```diff
@@ -0,0 +1,73 @@
+    public void writeResponseHeaders() {
+        final Map<String, String> headers = responseHeader.getHeaders();
+        headers.forEach((key, value) -> {
+            try {
```
어차피 Throw가 되는데 반복문 내에서 잡을 필요가 있을까요 ?

shared-moon commented on src/main/java/webserver/response/CustomResponse.java:
```diff
@@ -0,0 +1,73 @@
+            try {
+                dos.writeBytes(createContext(key, value));
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```
예외 발생시에는 좀 더 의미가 있는 예외 클래스를 사용하고, 메세지를 넣어주는게 좋아요 ~

shared-moon commented on src/main/java/webserver/session/SessionManager.java:
```diff
@@ -0,0 +1,41 @@
+        return session;
+    }
+
+    public static Session findSession(final String id) {
```
테스트를 제외하면 내부에서만 쓰는 메서드인데 private으로 하는게 맞지 않을까요 ?

shared-moon reviewed (approved):
혜원님 안녕하세요!
2단계 미션 잘 진행 해 주셨네요 :)
구조에 대한 피드백을 좀 남겨 두었는데요, 3단계가 리팩토링 미션이니 한번 진행 해보시면 좋을 것 같아요!
이번 PR은 바로 머지하도록 하겠습니다.
고생 많으셨어요!
